### ISSUE TITLE
Model contract (vector_field, build_terms, integrate, apply_bcs, diagnose)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/somax#21]
## Summary

Define the shared interface that every somax model follows. This is the core contract that enables composability with diffrax, JAX transformations, and downstream consumers.

Parent: Sub-epic: Core API Framework

## Design

```python
class SomaxModel(eqx.Module):
    # Required fields
    params: eqx.Module
    consts: eqx.Module
    grid: ...                          # finitevolX grid (or None for ODEs)
    mask: ArakawaCGridMask | None      # first-class, always present
    bcs: BoundaryConditionSet | None   # first-class, always present

    # Required methods
    def vector_field(self, t, state, args=None) -> State:
        """Explicit RHS: dstate/dt. Compatible with diffrax.ODETerm."""

    def apply_boundary_conditions(self, state) -> State:
        """Enforce BCs. Called inside vector_field. Never skipped."""

    def build_terms(self) -> dfx.AbstractTerm:
        """Build diffrax term(s). Supports ODE, SDE, IMEX."""
        return dfx.ODETerm(self.vector_field)  # default

    def integrate(self, state0, t0, t1, dt, **kw) -> dfx.Solution:
        """Forward integration. Uses build_terms() for term construction."""

    def diagnose(self, state) -> Diagnostics:
        """Compute diagnostic fields on demand."""

    # Optional methods
    def diffusion_field(self, t, state, args=None):
        """Stochastic diffusion. Returns None if deterministic."""
        return None

    def implicit_field(self, t, state, args=None):
        """Implicit part of IMEX. Returns None if fully explicit."""
        return None
```

## Key Design Decisions

- **BCs are first-class**: `apply_boundary_conditions()` is always called inside `vector_field`
- **Masks are first-class**: passed to all finitevolX operators
- **`build_terms()`** controls diffrax term structure (ODE vs MultiTerm for SDE/IMEX)
- **`integrate()`** uses `build_terms()`, picks default solver based on term type
- **`diagnose()`** returns Diagnostics pytree, never stores on model
- **`diffusion_field`** returns None by default (opt-in stochastic)
- **`implicit_field`** returns None by default (opt-in IMEX)

## Consumer Patterns (must all work)

```python
# Forward sim
sol = model.integrate(state0, t0, t1, dt)

# Param estimation
grads = jax.grad(lambda p: cost(eqx.tree_at(lambda m: m.params, model, p)))(model.params)

# State estimation  
grads = jax.grad(lambda s: cost(model.integrate(s, t0, t1, dt)))(state0)

# Ensemble vmap
results = jax.vmap(lambda s: model.integrate(s, t0, t1, dt).ys)(ensemble_states)
```

## Package Location

`src/somax/_src/models/base.py` or as a protocol

## Acceptance Criteria

- [ ] Base model class/protocol defined
- [ ] `apply_boundary_conditions` called inside default `vector_field` pattern
- [ ] `build_terms()` handles ODE, SDE (MultiTerm with ControlTerm), IMEX (MultiTerm)
- [ ] `integrate()` works with all term types, picks default solver
- [ ] All four consumer patterns verified (forward, param est, state est, vmap)
- [ ] Tests for `jax.grad` through `integrate` w.r.t. params and state
